### PR TITLE
Bindings: Query type kinds, derived types, and elements

### PR DESCRIPTION
**Pull request by tbennun**
This PR refactors TypeRef bindings out of `value.py` and enables querying types and their attributes. This includes the type kinds, derived types (e.g., the members of a struct), and element counts in arrays and vectors. Without this PR, obtaining this information could only be done via error-prone string processing.

### CODE REVIEW

File: llvmlite/binding/typeref.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+        return ffi.lib.LLVMPY_TypeIsVector(self)
+
+    @property
+    def is_function_vararg(self):
+        """
+        Returns true if a function type accepts a variable number of arguments.
+        When the type is not a function, raises exception.
+        """
+        if self.type_kind != TypeKind.function:
+            raise ValueError("Type {} is not a function".format(self))
```
Comment (sklam) [inline llvmlite/binding/typeref.py:77]: Can you add a test to exercise the exception raising?

File: llvmlite/binding/typeref.py [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+    @property
+    def element_count(self):
+        """
+        Returns the number of elements in an array or a vector. For scalable
+        vectors, returns minimum number of elements. When the type is neither
+        an array nor a vector, raises exception.
+        """
+        if not self.is_array and not self.is_vector:
+            raise ValueError("Type {} is not an array nor vector".format(self))
```
Comment (sklam) [inline llvmlite/binding/typeref.py:105]: Can you add a test to exercise the exception raising?

**Review by sklam — COMMENTED** [review]
Overall the PR looks good, it just need one minor change to increase the test coverage. Also, the PR currently has merge conflict in one of the Makefile.

**Comment by tbennun** [comment]
Thank you! I was planning to get to it later this week.

**Review by stuartarchibald — APPROVED** [review]
Thanks for the patch @tbennun and thanks for the review and additional fixes @sklam. As @sklam has already reviewed fb025b6..e1bf86f and has indicted approval conditional on an additional test being written (https://github.com/numba/llvmlite/pull/978#pullrequestreview-1766002637) I've reviewed and am approving the merge from `main` in fe53feb and the test addition in 80bb55b. Great to see this feature implemented, thanks again all!

**Pull request merged**